src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceBatchSource.java

```diff
@@ -0,0 +1,264 @@
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(SalesforceBatchSource.NAME)
+@Description("Plugin to read data from Salesforce in batches.")
+public class SalesforceBatchSource extends BatchSource<String, String, StructuredRecord> {
+  static final String NAME = "SalesforceBatchSource";
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceBatchSource.class);
+
+  private final Config config;
+  private Schema schema;
+
+  SalesforceBatchSource(Config config) throws ConnectionException {
+    this.config = config;
+  }
+
+  static final class Config extends BaseSalesforceConfig {
+    private static final String PROPERTY_QUERY = "query";
+
+    @Description("The SOQL query to retrieve results from")
+    @Macro
+    private final String query;
+
+    Config(Configuration conf) {
+        super(null,
+              conf.get(SalesforceConstants.CLIENT_ID), conf.get(SalesforceConstants.CLIENT_SECRET),
+              conf.get(SalesforceConstants.USERNAME), conf.get(SalesforceConstants.PASSWORD),
+              conf.get(SalesforceConstants.LOGIN_URL));
+
+        this.query = conf.get(SalesforceConstants.QUERY);
+    }
+
+    public String getQuery() {
+      return query;
+    }
+
+    @Override
+    public void validate() {
+      super.validate();
+
+      if (!containsMacro(PROPERTY_QUERY)) {
+        SalesforceQueryParser.validateQuery(query);
+      }
+    }
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws ConnectionException {
+    this.schema = SalesforceSchemaUtil.getSchemaFromQuery(config.getAuthenticatorCredentials(), config.getQuery());
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    config.validate(); // validate when macros not yet substituted
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(schema);
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws ConnectionException {
+    config.validate(); // validate when macros are already substituted
+
+    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
+    lineageRecorder.createExternalDataset(schema);
+
+    context.setInput(Input.of(config.referenceName, new SalesforceInputFormatProvider(config)));
+
+    if (schema != null) {
+      if (schema.getFields() != null) {
+        lineageRecorder.recordRead("Read", "Read from Salesforce",
+                                   schema.getFields().stream().map(Schema.Field::getName)
+                                     .collect(Collectors.toList()));
+      }
+    }
+  }
+
+  @Override
+  public void transform(KeyValue<String, String> input,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+    String[] fieldNames = getValuesFromCSVRow(input.getKey());
+    String[] values = getValuesFromCSVRow(input.getValue());
+
+    if (fieldNames.length != values.length) {
```

Comment:
Can we emit the record to ErrorCollector here with error code and message instead of throwing exception? Additional you would have to enable error port in widget JSON.  Check data-integrations/wrangler/wrangler-transform plugin on how to set it.

Reply:
at that point there the record is not yet built since the csv was invalid. But emitter.emitError needs to send structedRecord. Can you please explain what should be sent in this case?

Reply:
done